flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/Knn.java

```diff
@@ -0,0 +1,211 @@
+                                    featureDim);
+                        }
+                        DenseVector featureNorms = computeNorm(packedFeatures);
+                        if (dataPoints.size() > 0) {
+                            out.collect(new KnnModelData(packedFeatures, featureNorms, labels));
+                        }
+                    }
+                });
+    }
+
+    /**
+     * Merges knn model data.
+     *
+     * @param distributedModelData Distributed knn model data.
+     * @return Knn model.
+     */
+    private static DataStream<KnnModelData> mergeModelData(
+            DataStream<KnnModelData> distributedModelData) {
+        distributedModelData.getTransformation().setParallelism(1);
+        return DataStreamUtils.mapPartition(
+                distributedModelData,
+                new RichMapPartitionFunction<KnnModelData, KnnModelData>() {
+                    @Override
+                    public void mapPartition(
+                            Iterable<KnnModelData> values, Collector<KnnModelData> out) {
+                        List<KnnModelData> buffer = new ArrayList<>(1);
+                        int totalNumVec = 0;
+                        for (KnnModelData data : values) {
+                            totalNumVec += data.packedFeatures.numCols();
+                            buffer.add(data);
+                        }
+                        int featureDim = buffer.get(0).packedFeatures.numRows();
+                        DenseMatrix packedFeatures = new DenseMatrix(featureDim, totalNumVec);
+                        DenseVector featureNorms = new DenseVector(totalNumVec);
+                        DenseVector labels = new DenseVector(totalNumVec);
+                        int offset = 0;
+                        for (KnnModelData data : buffer) {
+                            int numVec = data.featureNorms.size();
+                            System.arraycopy(
+                                    data.packedFeatures.values,
+                                    0,
+                                    packedFeatures.values,
+                                    offset * featureDim,
+                                    featureDim * numVec);
+                            System.arraycopy(
+                                    data.featureNorms.values,
+                                    0,
+                                    featureNorms.values,
+                                    offset,
+                                    numVec);
+                            System.arraycopy(data.labels.values, 0, labels.values, offset, numVec);
+                            offset += numVec;
+                        }
+                        out.collect(new KnnModelData(packedFeatures, featureNorms, labels));
+                    }
+                });
+    }
+
+    /**
+     * For Euclidean distance, distance = sqrt((a - b)^2) = (sqrt(a^2 + b^2 - 2ab)) So it can
```

Comment:
Can we polish the comments here and move it to `prepareModelData()`?

Reply:
OK